cpp/tests/iterator/iterator_test.cu

```diff
@@ -0,0 +1,393 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* Proof the concept of iterator driven aggregations to reuse the logic
+
+   The concepts:
+   1. computes the aggregation by given iterators
+   2. computes by using cub and thrust with same function parameters
+   3. accepts nulls and group_by with same function parameters
+
+    CUB reduction:  https://nvlabs.github.io/cub/structcub_1_1_device_reduce.html#aa4adabeb841b852a7a5ecf4f99a2daeb
+    Thrust reduction: https://thrust.github.io/doc/group__reductions.html#ga43eea9a000f912716189687306884fc7
+
+    Thrust iterators: https://thrust.github.io/doc/group__fancyiterator.html
+*/
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <bitset>
+#include <cstdint>
+#include <iostream>
+#include <numeric>
+#include <iterator>
+
+#include <utilities/cudf_utils.h> // need for CUDA_HOST_DEVICE_CALLABLE
+#include <utilities/device_atomics.cuh> // need for device operators.
+
+#include <tests/utilities/column_wrapper.cuh>
+#include <tests/utilities/cudf_test_fixtures.h>
+
+#include <cub/device/device_reduce.cuh>
+#include <thrust/device_vector.h>
+#include <thrust/transform.h>
+
+#include <thrust/tuple.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/iterator/iterator_adaptor.h>
+
+
+// --------------------------------------------------------------------------------------------------------
+// structs for column_input_iterator
+template<typename T>
+struct ColumnDataNonNull{
+    const T *data;
+
+    __device__ __host__
+    ColumnDataNonNull(T *_data)
+    : data(_data){};
+
+    __device__ __host__
+    T at(gdf_index_type id) const {
+        return data[id];
+    };
+};
+
+template<typename T>
+struct ColumnData{
```

Comment:
In a production version of this design, instead of having both `ColumnDataNonNull` and `ColumnData`, I'd add a `bool` template argument to `ColumnData` indicating whether or not the `valid` mask is present. In that way, you can avoid the performance hit from always checking `gdf_is_valid` by checking `if(nulls_present)` if there aren't any nulls in the column. Since it's a template argument, and known at compile time, the compiler can optimize out the `if` check if `nulls_present == false`.

Reply:
Those 2 structs are different in the count of the members.
I'm not sure if I set bool template argument `nulls_present`, the compiler optimized out the unused struct member if `nulls_present` = false?
If the member is not optimized, the copy of iterator may take unnecessary copies.

Reply:
You could explicitly specialize for `true` and `false` and then you can have different members of the class depending on the template arg.

Reply:
That sounds good/

Reply:
That said, I don't know if it's necessary. Having an unused 8B pointer member of the iterator probably isn't going to have much overhead, is it?

Reply:
This is addressed in latest commit. Please check.